kyo-core/shared/src/main/scala/kyo/Queue.scala

```diff
@@ -210,118 +352,38 @@ object Queue:
-                            val u = q.unsafe
-                            if u.offer(v) then ()
-                            else
-                                discard(u.poll())
-                                loop(v)
-                            end if
-                        end loop
-                        loop(v)
-                        true
-                    end offer
-                    def poll()(using AllowUnsafe) = u.poll()
-                    def peek()(using AllowUnsafe) = u.peek()
-            )
-        }
-
-    private def fromJava[A](q: java.util.Queue[A])(using frame: Frame): Unbounded[A] =
-        new Unbounded(
-            frame,
-            new Unsafe[A]:
-                def capacity                       = Int.MaxValue
-                def size()(using AllowUnsafe)      = q.size
-                def empty()(using AllowUnsafe)     = q.isEmpty()
-                def full()(using AllowUnsafe)      = false
-                def offer(v: A)(using AllowUnsafe) = q.offer(v)
-                def poll()(using AllowUnsafe)      = Maybe(q.poll)
-                def peek()(using AllowUnsafe)      = Maybe(q.peek)
-        )
-
-    private def fromJava[A](q: java.util.Queue[A], _capacity: Int)(using frame: Frame): Queue[A] =
-        new Queue(
-            frame,
-            new Unsafe[A]:
-                def capacity                       = _capacity
-                def size()(using AllowUnsafe)      = q.size
-                def empty()(using AllowUnsafe)     = q.isEmpty()
-                def full()(using AllowUnsafe)      = q.size >= _capacity
-                def offer(v: A)(using AllowUnsafe) = q.offer(v)
-                def poll()(using AllowUnsafe)      = Maybe(q.poll)
-                def peek()(using AllowUnsafe)      = Maybe(q.peek)
-        )
+        def fromJava[A](q: java.util.Queue[A], _capacity: Int = Int.MaxValue)(using initFrame: Frame, allow: AllowUnsafe): Unsafe[A] =
+            new Closeable[A](initFrame):
+                def capacity                   = _capacity
+                def size()(using AllowUnsafe)  = op(q.size())
+                def empty()(using AllowUnsafe) = op(q.isEmpty())
+                def full()(using AllowUnsafe)  = op(q.size() >= _capacity)
+                def offer(v: A)(using AllowUnsafe) =
+                    offerOp(
+                        q.offer(v),
+                        try !q.remove(v)
+                        catch
+                            case _: UnsupportedOperationException =>
+                                // TODO the race repair should use '!q.remove(v)' but JCTools doesn't support the operation.
+                                // In rare cases, items may be left in the queue permanently after closing due to this limitation.
+                                // The item will only be removed when the queue object itself is garbage collected.
+                                !q.contains(v)
+                    )
```

Comment:
This is the repair logic for the race condition. There's an issue with JCTools not supporting `remove` as described in the comment. It seems reasonable since the race condition should be quite rare.